examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,605 @@
+per epoch using a P100 GPU accelerator"""
+
+print(f"Number of examples (caption-image pairs): {train_example_count}")
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {train_example_count // batch_size}")
+train_data = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+history = dual_encoder.fit(train_data, epochs=num_epochs)
+
+"""
+Plotting the training loss:
+"""
+
+plt.plot(history.history["loss"])
+plt.xlabel("Epochs")
+plt.ylabel("Loss")
+plt.show()
+
+"""
+## Search for images using natural language
+
+The process of searching for images using using natural language consists of the
+following steps:
+1. Generating embeddings for the images by feeding them into the `vision_encoder`.
+2. Feed the natural language query to the `text_encoder` to generate a query embedding.
+3. Compute the similarity between the query embedding and the image embeddings
+in the index to retrieve the indices of the top matches.
+4. Lookup the paths of the top matching images to display them.
+
+Note that, after training the `dual encoder`, the only the fine-tuned `vision_encoder`
+and `text_encoder` models will be used, while the `dual_encoder` model will be discarded.
+"""
+
+"""
+### Generate embeddings for the images
+
+We load the images and feed them into the `vision_encoder` to generate their embeddings.
+In large scale systems, this step is performed using a parallel data processing frameworks,
+such as [Apache Spark](https://spark.apache.org) or [Apache Beam](https://beam.apache.org).
+Generating the image embeddings will take a few minutes.
+"""
+
+image_embeddings = []
+idx = 0
+batch_size = 512
+
+print(f"Image embeddings generation started...")
+print(f"Generating embeddings for {len(image_paths)} images...")
+while idx < len(image_paths):
+    image_batch = []
+    current_image_paths = image_paths[idx : idx + batch_size]
+    for image_path in current_image_paths:
+        raw_image = tf.io.read_file(image_path)
+        image_array = tf.image.decode_jpeg(raw_image, channels=3)
+        image_array = tf.image.resize(image_array, (299, 299))
+        image_batch.append(image_array)
+    idx += batch_size
+    current_image_embeddings = vision_encoder(image_batch)
+    image_embeddings.extend(current_image_embeddings)
+    if idx % 10 == 0:
+        print(f"Generated embeddings for {idx} images...")
+
+image_embeddings = np.array(image_embeddings)
+print(f"Image embeddings generation completed.")
+print(f"Image embeddings shape: {image_embeddings.shape}.")
+
+"""
+### Retrieve relevant images
+
+In this example, we use exact matching by computing the dot product similarity
+between the input query embedding and the image embeddings, and retrieve the top k
+matches. However, *approximate* similarity matching, using frameworks like
+[ScaNN](https://github.com/google-research/google-research/tree/master/scann),
+[Annoy](https://github.com/spotify/annoy), or [Faiss](https://github.com/facebookresearch/faiss)
```

Comment:
Fun fact: "faiss" would be pronounced like "fesse" (ass cheek) in French.

Reply:
Hahaha, I know!!! (I speak french a little bit as my wife is French).